test/release/examples/primers/forallLoops.chpl

```diff
@@ -0,0 +1,305 @@
+*/
+
+var D = [(a,b,c) in zip(A,B,C)] a + c - b;
+
+writeln("The result of may-parallel expression, D is:");
+writeln(D);
+writeln();
+
+/*
+As with must-parallel loops, if A were a distributed array,
+iteration locality while computing the above expression
+would be determined by A's domain map.
+
+Task Intents and Shadow Variables
+---------------------------------
+
+A forall loop may name some variables declared outside the loop,
+or "outer variables". If so, "shadow variables" are introduced.
+Each task created by the parallel iterator gets its own set of
+shadow variables, one per outer variable.
+
+ - Each shadow variable behaves as if it were a formal argument
+   of the task function for the task. The outer variable is passed
+   to this formal argument according to the argument intent
+   associated with the shadow variable, which is called a "task intent".
+
+ - The name of an outer variable within the lexical scope of the loop
+   body refers implicitly to the corresonding shadow variable. If the
+   parallel iterator causes multiple iterations of the loop to be
+   executed by the same task, these iterations refer to the same set
+   of shadow variables.
+
+ - Each shadow variable is deallocated at the end of its task.
+
+The default argument intent is used by default. For numeric types,
+this implies capturing the value of the outer variable by the time
+the task starts executing. Arrays are passed by reference, and so are
+sync, single, and atomic variables.
+*/
+
+var outerIntVariable = 1;
+proc updateOuterVariable() {
+  outerIntVariable += 1;  // always refers to the outer variable
+}
+var outerAtomicVariable: atomic int;
+
+forall i in 1..n {
+  if i == 1 then
+    updateOuterVariable();     // beware of potential for data races
+
+  // the shadow variable always contains the value as of loop start
+  writeln("shadow outerIntVariable is: ", outerIntVariable);
+
+  D[i] += 0.5;                 // beware of potential for data races
+
+  outerAtomicVariable.add(3);  // ok: concurrent updates are atomic
+}
+
+writeln();
+writeln("After a loop with default intents, D is:");
+writeln(D);
+writeln("outerIntVariable is: ", outerIntVariable);
+writeln("outerAtomicVariable is: ", outerAtomicVariable.read());
+writeln();
+
+/*
+The task intents ``in``, ``const in``, ``ref``, ``const ref``,
+and ``reduce`` can be specified explicitly using a ``with`` clause.
+
+An ``in`` or ``const in`` intent creates a copy of the outer variable
+for each task. A ``ref`` or ``const ref`` makes the
+shadow variable an aliass for the outer variable.
+*/
+
+var outerRealVariable = 1.0;
+
+forall i in 1..n with (in outerIntVariable,
+                       ref outerRealVariable) {
+  outerIntVariable += 1; // a per-task copy, never accessed concurrently
+
+  if i == 1 then
+    outerRealVariable *= 2;  // beware of potential for data races
+}
+
+writeln("After a loop with explicit intents:");
+writeln("outerIntVariable is: ", outerIntVariable);
```

Comment:
I think it would be helpful to call out that this value is *the same as it was before the loop* because of the "in" intent, to explicitly drive home what happened and why it was important.